File: java/L100/LeetCode0146.java
package L100;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class LeetCode0146 {
    public static void main(String[] args) {
        LRUCache2 cache2 = new LRUCache2(2);
        cache2.put(1, 1);
        cache2.put(2, 2);
        cache2.put(3, 3);
        System.out.println(cache2.get(1));
        System.out.println(cache2.get(2));
    }
}

class LRUCache {
    class DNode {
        int key;
        int value;
        DNode prev, next;
        public DNode(int k, int v) {
            key = k;
            value = v;
        }
    }
    private int cap;
    private DNode head;
    private Map<Integer,DNode> cache;

    public LRUCache(int capacity) {
        cache = new HashMap<>();
        head = new DNode(0, 0);
        head.next = head;
        head.prev = head;
        cap = capacity;
    }

    public void addToHead(DNode cur) {
        cur.next = head.next;
        head.next.prev = cur;
        head.next = cur;
        cur.prev = head;
    }

    public void removeNode(DNode cur) {
        cur.next.prev = cur.prev;
        cur.prev.next = cur.next;
    }
    
    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }
        DNode cur = cache.get(key);
        removeNode(cur);
        addToHead(cur);
        return cur.value;
    }
    
    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            DNode cur = cache.get(key);
            cur.value = value;
            removeNode(cur);
            addToHead(cur);
        } else if (cache.size() < cap) {
            DNode cur = new DNode(key,value);
            cache.put(key, cur);
            addToHead(cur);
        } else {
            DNode cur = head.prev;
            cache.remove(cur.key);
            head.prev = cur.prev;
            cur.prev.next = head;
            cur.key = key;
            cur.value = value;
            addToHead(cur);
            cache.put(key, cur);
        }
    }
}

class LRUCache2 extends LinkedHashMap<Integer, Integer> {
    private int capacity;

    public LRUCache2(int capacity) {
        super(capacity, 0.75F, true);
        this.capacity = capacity;
    }

    public Integer get(Integer key) {
        return super.getOrDefault(key, -1);
    }

    @Override
    protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
        return size() > capacity;
    }
}

File: cpp/L100/LeetCode0146.cc
#include <unordered_map>
struct Node {
    int key, value;
    Node *prev, *next;

    Node(int key, int value): key(key), value(value), prev(nullptr), next(nullptr) {}
};

class LRUCache {
    std::unordered_map<int, Node*> map;
    Node head;
    Node *tail;
    int size, used;

    void pushTail(Node *cur) {
        tail->next = cur;
        cur->prev = tail;
        tail = cur;
    }

    void removeFirst() {
        if (head.next == nullptr) {
            return;
        }
        Node *cur = head.next;
        if (cur == nullptr) {
            return;
        }
        head.next = cur->next;
        if (cur->next != nullptr) {
            cur->next->prev = &head;
        } else {
            tail = &head;
        }
        map.erase(cur->key);
        delete cur;
        --used;
    }
public:
    LRUCache(int capacity): head(0 ,0), size(capacity), used(0) {
        tail = &head;
    }
    
    int get(int key) {
        auto it{map.find(key)};
        if (it == map.end()) {
            return -1;
        }
        Node *cur = it->second;
        if (cur != tail) {
            cur->prev->next = cur->next;
            cur->next->prev = cur->prev;
            pushTail(cur);
        }
        return cur->value;
    }
    
    void put(int key, int value) {
        auto it{map.find(key)};
        if (it == map.end()) {
            Node *cur = new Node(key, value);
            if (used + 1 > size) {
                removeFirst();
            }
            map[key] = cur;
            pushTail(cur);
            ++used;
        } else {
            Node *cur = it->second;
            cur->value = value;
            if (cur != tail) {
                cur->prev->next = cur->next;
                cur->next->prev = cur->prev;
                pushTail(cur);
            }
        }
    }
};